proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,172 @@
+- if the lambda return type is explicit, that type is used;
+- if the lambda has no return expressions, the return type is `void` or `System.Threading.Tasks.Task` if `async`;
+- if the common type from the natural type of all `return` expressions in the body is the type `R0`, the return type is `R0` or `System.Threading.Tasks.Task<R0>` if `async`.
+
+A method group has a natural type if the method group contains a single method and the method has no unbound type parameters.
+
+The delegate type for the lambda or method group and parameter types `P1, ..., Pn` and return type `R` is:
+- if any parameter or return value is not by value, or there are more than 16 parameters, or any of the parameter types or return are not valid type arguments (say, `(int* p) => { }`), then the delegate is a synthesized `internal` anonymous delegate type with signature that matches the lambda or method group, and with parameter names `arg1, ..., argn` or `arg` if a single parameter;
+- if `R` is `void`, then the delegate type is `System.Action<P1, ..., Pn>`;
+- otherwise the delegate type is `System.Func<P1, ..., Pn, R>`.
+
+`modopt()` or `modreq()` in the method group signature are ignored in the corresponding delegate type.
+
+If synthesized delegate types are required, the compiler will attempt to reuse delegate types across multiple use sites. The compiler could generate generic delegate types parameterized by parameter types and return type similar to the generic types synthesized for anonymous types. But reuse might be limited to simply reusing delegate types when lambda or method group signatures match exactly.
+
+The anonymous delegate types will not be co- or contra-variant unlike the delegates constructed from `System.Action<>` and `System.Func<>`.
+
+Lambdas or method groups with natural types can be used as initializers in `var` declarations.
+
+```csharp
+var f1 = () => default;        // error: no natural type
+var f2 = x => { };             // error: no natural type
+var f3 = x => x;               // error: no natural type
+var f4 = () => 1;              // System.Func<int>
+var f5 = () : string => null;  // System.Func<string>
+```
+
+```csharp
+static void F1() { }
+static void F1<T>(this T t) { }
+static void F2(this string s) { }
+
+var f6 = F1;    // error: multiple methods
+var f7 = "".F1; // System.Action
+var f8 = F2;    // System.Action<string> 
+```
+
+### Implicit conversion to `System.Delegate`
+A consequence of inferring a natural type is that lambda expressions and method groups with natural type are implicitly convertible to `System.Delegate`.
+```csharp
+static void Invoke(Func<string> f) { }
+static void Invoke(Delegate d) { }
+
+static string GetString() => "";
+static int GetInt() => 0;
+
+Invoke(() => "");  // Invoke(Func<string>)
+Invoke(() => 0);   // Invoke(Delegate) [new]
+
+Invoke(GetString); // Invoke(Func<string>)
+Invoke(GetInt);    // Invoke(Delegate) [new]
+```
+
```

Comment:
Think we need a couple of counter examples here showing where method group and lambda conversions will not work after this change. Specifically:

1. Cases where there are multiple candidates in the method group which means we can't infer a natural type
2. Cases where the return type of the lambda cannot be inferred hence the lambda does not have a natural type